test/e2e/framework/pod/wait.go

```diff
@@ -197,43 +235,91 @@ func WaitForPodsRunningReady(c clientset.Interface, ns string, minPods, allowedN
 		return false, nil
 	}) != nil {
 		if !ignoreNotReady {
-			return errors.New(errorBadPodsStates(badPods, desiredPods, ns, "RUNNING and READY", timeout, lastAPIError))
+			return errorBadPodsStates(badPods, desiredPods, ns, "RUNNING and READY", timeout, lastAPIError)
 		}
 		e2elog.Logf("Number of not-ready pods (%d) is below the allowed threshold (%d).", notReady, allowedNotReadyPods)
 	}
 	return nil
 }
 
 // WaitForPodCondition waits a pods to be matched to the given condition.
-func WaitForPodCondition(c clientset.Interface, ns, podName, desc string, timeout time.Duration, condition podCondition) error {
-	e2elog.Logf("Waiting up to %v for pod %q in namespace %q to be %q", timeout, podName, ns, desc)
-	var lastPodError error
-	for start := time.Now(); time.Since(start) < timeout; time.Sleep(poll) {
+func WaitForPodCondition(c clientset.Interface, ns, podName, conditionDesc string, timeout time.Duration, condition podCondition) error {
+	e2elog.Logf("Waiting up to %v for pod %q in namespace %q to be %q", timeout, podName, ns, conditionDesc)
+	var (
+		lastPodError error
+		lastPod      *v1.Pod
+		start        = time.Now()
+	)
+	err := wait.PollImmediate(poll, timeout, func() (bool, error) {
 		pod, err := c.CoreV1().Pods(ns).Get(context.TODO(), podName, metav1.GetOptions{})
 		lastPodError = err
 		if err != nil {
-			if apierrors.IsNotFound(err) {
-				e2elog.Logf("Pod %q in namespace %q not found. Error: %v", podName, ns, err)
-			} else {
-				e2elog.Logf("Get pod %q in namespace %q failed, ignoring for %v. Error: %v", podName, ns, poll, err)
-			}
-			continue
+			return handleWaitingAPIError(err, true, "getting pod %s", podIdentifier(ns, podName))
 		}
+		lastPod = pod // Don't overwrite if an error occurs after successfully retrieving.
+
 		// log now so that current pod info is reported before calling `condition()`
 		e2elog.Logf("Pod %q: Phase=%q, Reason=%q, readiness=%t. Elapsed: %v",
 			podName, pod.Status.Phase, pod.Status.Reason, podutils.IsPodReady(pod), time.Since(start))
 		if done, err := condition(pod); done {
 			if err == nil {
-				e2elog.Logf("Pod %q satisfied condition %q", podName, desc)
+				e2elog.Logf("Pod %q satisfied condition %q", podName, conditionDesc)
 			}
-			return err
+			return true, err
+		} else if err != nil {
+			// TODO(#109732): stop polling and return the error in this case.
+			e2elog.Logf("Error evaluating pod condition %s: %v", conditionDesc, err)
 		}
+		return false, nil
+	})
+	if err == nil {
+		return nil
+	}
+	if IsTimeout(err) {
+		e2elog.Logf("Timed out while waiting for pod %s to be %s. Last observed as: %s",
+			podIdentifier(ns, podName), conditionDesc, format.Object(lastPod, 1))
 	}
-	if apierrors.IsNotFound(lastPodError) {
-		// return for compatbility with other functions testing for IsNotFound
-		return lastPodError
+	if lastPodError != nil {
+		// If the last API call was an error.
+		err = lastPodError
```

Comment:
@tallclair: was removal of the `apierrors.IsNotFound` intentional?

It looked like it was, but I am not sure anymore:
- The "If the last API call was an error." comment is incomplete.
- The example with `framework.ExpectNoError(e2epod.WaitForPodNotPending(f.ClientSet, f.Namespace.Name, "does-not-exist"))` now times out after five minutes, which is not what you showed earlier.

Reply:
FWIW, I think treating "Not Found" as a transient error makes sense because the WaitFor functions become more versatile.

Reply:
Not Found can mean "pod already existed and was deleted" or "pod was not created yet"
I would not blindly rely on that, at least not expose it to users since one of those uses may be racy for the test

Reply:
> was removal of the apierrors.IsNotFound intentional?

Yes, in the event of a timeout, the last API error is always returned now. In other words, the behavior is unchanged for the NotFound case, but behaves the same way for other API errors. I think a better way of handling this would be to have the timeout error wrap the API error (implement an Unwrap) function, so it can be recognized as both a timeout and an API error.

> If the last API call was an error.

I actually think there's a bug here, it should only override timeout errors (or not, see above), but it shouldn't override any other type of error. I'm working on a follow-up that integrates this change, but I could also push it on it's own if needed.

WIP: https://github.com/kubernetes/kubernetes/compare/master...tallclair:wait?expand=1
Specifically, see https://github.com/tallclair/kubernetes/blob/1414ef0608e356039ee1b1ddaf459b221b7010d2/test/e2e/framework/wait/wait.go#L93-L109

> The example with framework.ExpectNoError(e2epod.WaitForPodNotPending(f.ClientSet, f.Namespace.Name, "does-not-exist")) now times out after five minutes, which is not what you showed earlier.

I artificially changed the timeout to 20 seconds in the example I pasted - is that what you mean? That was just lowering the timeout time. I'm working on exposing the NotFound behavior to the caller - see https://github.com/tallclair/kubernetes/blob/1414ef0608e356039ee1b1ddaf459b221b7010d2/test/e2e/framework/wait/wait.go#L41-L42

Reply:
You are right, it was also timing out in the previous comment.